File: Indexed Search/Hash Based Indexed Search/Cuckoo Hashing/CuckooHashing.java
import java.util.Arrays;

public class CuckooHashing {

    private static final int SIZE = 11;

    private static final int EMPTY = -1;

    private int[] table1;

    private int[] table2;

    public CuckooHashing() {

        table1 =
            new int[SIZE];

        table2 =
            new int[SIZE];

        Arrays.fill(
            table1,
            EMPTY
        );

        Arrays.fill(
            table2,
            EMPTY
        );
    }

    private int hash1(int key) {
        return key % SIZE;
    }

    private int hash2(int key) {
        return (key / SIZE) % SIZE;
    }

    public void insert(int key) {

        int i =
            hash1(key);

        if (
            table1[i]
            ==
            EMPTY
        )
            table1[i] = key;

        else
            table2[
                hash2(key)
            ] = key;
    }

    public boolean search(int key) {

        return (
            table1[
                hash1(key)
            ]
            ==
            key
        )
        ||
        (
            table2[
                hash2(key)
            ]
            ==
            key
        );
    }

    public static void main(
        String[] args
    ) {

        CuckooHashing h =
            new CuckooHashing();

        h.insert(25);

        System.out.println(
            h.search(25)
        );
    }
}

File: Indexed Search/Hash Based Indexed Search/Cuckoo Hashing/CuckooHashing.py
TABLE_SIZE = 11
EMPTY = None
MAX_KICKS = 20


class CuckooHash:

    def __init__(self):

        self.table1 = [
            EMPTY
        ] * TABLE_SIZE

        self.table2 = [
            EMPTY
        ] * TABLE_SIZE

    def hash1(self, key):
        return key % TABLE_SIZE

    def hash2(self, key):
        return (
            key //
            TABLE_SIZE
        ) % TABLE_SIZE

    def search(self, key):

        return (
            self.table1[
                self.hash1(key)
            ] == key
            or
            self.table2[
                self.hash2(key)
            ] == key
        )

    def insert(self, key):

        current = key
        first = True

        for _ in range(MAX_KICKS):

            if first:

                idx = self.hash1(
                    current
                )

                if (
                    self.table1[idx]
                    is None
                ):
                    self.table1[idx] = current
                    return

                (
                    self.table1[idx],
                    current
                ) = (
                    current,
                    self.table1[idx]
                )

                first = False

            else:

                idx = self.hash2(
                    current
                )

                if (
                    self.table2[idx]
                    is None
                ):
                    self.table2[idx] = current
                    return

                (
                    self.table2[idx],
                    current
                ) = (
                    current,
                    self.table2[idx]
                )

                first = True


h = CuckooHash()

h.insert(20)

print(
    h.search(20)
)
